File: lib/voice-engine/ugc-schema.ts
// PR Sprint 7.18 — UGCBundle schema port.
//
// 1:1 translation of
// Helm SEO/helm-adaptive-voice-engine/ugc_schema.py.
// Pydantic v2 → Zod. Same field types, same length / range
// constraints, same custom validators (hook word count,
// overlay word count, body beats sequential, hashtag format).
//
// Bundle shape stays exactly the same as the Python source so a
// Python service or a TS service could interchange JSON payloads
// without translation.

import { z } from 'zod';

// ============================================================
// Word-counting helper
// ============================================================
//
// Hotfix — the original word counters did
//   `text.trim().split(/\s+/).length`
// which counts EVERY whitespace-separated token, including
// standalone emoji like ❌ or ✅. That punished a deliberately
// good overlay pattern ("BUFFER ❌ NOTION ❌ FIGMA ❌") — three
// brand names plus three pictogram markers, the exact "brand-tool
// callout" shape ugc_validator.py recommends. Splitting that on
// whitespace gives 6 tokens, the schema rejected it as "6 words",
// and video generation failed for any prompt that produced a
// brand-vs-brand comparison overlay.
//
// `countLexicalWords` only counts tokens that contain at least
// one Unicode letter or number. Pictogram-only tokens (emoji,
// dashes, bullets, arrows) don't increment the count.
//
//   "BUFFER ❌ NOTION ❌ FIGMA ❌"  → 3 (was 6)
//   "I dropped Buffer last month"  → 5 (unchanged)
//   "Stop. Now."                   → 2 (unchanged)
//   "$1M in 6 months"              → 4 (unchanged — digits count)
//   "❌"                            → 0 (was 1; pure pictograms aren't
//                                       a word)
//
// Same logic mirrored in lib/voice-engine/ugc_schema.py so the
// Python and TS paths agree.
function countLexicalWords(text: string): number {
  return text
    .trim()
    .split(/\s+/)
    .filter((token) => token.length > 0 && /[\p{L}\p{N}]/u.test(token))
    .length;
}

// ============================================================
// Enums
// ============================================================

export const DELIVERY_STYLES = [
  'punchy', // Fast, high-energy, attention-grabbing
  'explanatory', // Calm, informative
  'tension', // Building toward a reveal, slightly slower
  'reveal', // The payoff moment, emphatic but controlled
  'warm', // Friendly, conversational close
  'confessional', // First-person admission, slightly vulnerable
  'emphatic', // Stressed, slow. Use sparingly.
] as const;

export const DeliveryStyleSchema = z.enum(DELIVERY_STYLES);
export type DeliveryStyle = z.infer<typeof DeliveryStyleSchema>;

// ============================================================
// Sections
// ============================================================

export const HookSectionSchema = z
  .object({
    text: z
      .string()
      .min(10, 'Hook text must be at least 10 chars')
      .max(180, 'Hook text must be at most 180 chars'),
    duration_seconds: z.number().min(1.0).max(4.0),
    delivery: DeliveryStyleSchema,
  })
  .strict()
  .superRefine((hook, ctx) => {
    // 9-word cap, same as the Pydantic field_validator. Uses the
    // lexical counter so a hook like "I quit 3 apps. ✋" is 4
    // words, not 5.
    const wordCount = countLexicalWords(hook.text);
    if (wordCount > 9) {
      ctx.addIssue({
        code: 'custom',
        message: `Hook has ${wordCount} words. Maximum is 9 spoken words. Current hook: '${hook.text}'. Trim it aggressively.`,
        path: ['text'],
      });
    }
  });

export type HookSection = z.infer<typeof HookSectionSchema>;

export const BodyBeatSchema = z
  .object({
    beat: z.number().int().min(1).max(5),
    text: z
      .string()
      .min(20, 'Body beat text must be at least 20 chars')
      .max(400, 'Body beat text must be at most 400 chars'),
    duration_seconds: z.number().min(2.0).max(15.0),
    delivery: DeliveryStyleSchema,
  })
  .strict();

export type BodyBeat = z.infer<typeof BodyBeatSchema>;

export const CTASectionSchema = z
  .object({
    text: z
      .string()
      .min(10, 'CTA text must be at least 10 chars')
      .max(200, 'CTA text must be at most 200 chars'),
    duration_seconds: z.number().min(2.0).max(6.0),
    delivery: DeliveryStyleSchema,
  })
  .strict();

export type CTASection = z.infer<typeof CTASectionSchema>;

export const OverlaySchema = z
  .object({
    text: z
      .string()
      .min(1, 'Overlay text required')
      .max(40, 'Overlay text must be at most 40 chars'),
    trigger_at_seconds: z.number().min(0.0),
    duration_seconds: z.number().min(0.5).max(5.0),
  })
  .strict()
  .superRefine((overlay, ctx) => {
    // Lexical word count so emoji-only tokens (❌ ✅ → •) don't
    // inflate the count. ugc_validator.py explicitly recommends
    // the "BRAND ❌ BRAND ❌ BRAND ❌" shape — counting the ❌
    // as words made that shape un-shippable.
    const wordCount = countLexicalWords(overlay.text);
    if (wordCount > 5) {
      ctx.addIssue({
        code: 'custom',
        message: `Overlay '${overlay.text}' has ${wordCount} words; max is 5. Overlays longer than 5 words are an anti-pattern.`,
        path: ['text'],
      });
    }
  });

export type Overlay = z.infer<typeof OverlaySchema>;

export const UGCMetadataSchema = z
  .object({
    language: z.string().min(2).max(8).default('en'),
    platform: z.string(), // tiktok / instagram / threads / linkedin / facebook
    passes_swipe_test: z.boolean().default(true),
  })
  .strict();

export type UGCMetadata = z.infer<typeof UGCMetadataSchema>;

// ============================================================
// Top-level bundle
// ============================================================

export const UGCBundleSchema = z
  .object({
    hook: HookSectionSchema,
    body: z.array(BodyBeatSchema).min(1).max(5),
    cta: CTASectionSchema,
    overlays: z.array(OverlaySchema).min(3).max(8),
    caption: z
      .string()
      .min(20, 'Caption must be at least 20 chars')
      .max(500, 'Caption must be at most 500 chars'),
    hashtags: z.array(z.string()).min(3).max(5),
    metadata: UGCMetadataSchema,
  })
  .strict()
  .superRefine((bundle, ctx) => {
    // body beats must be sequential 1..N
    for (let i = 0; i < bundle.body.length; i++) {
      if (bundle.body[i].beat !== i + 1) {
        ctx.addIssue({
          code: 'custom',
          message: `Body beats must be sequential starting from 1 (got beat=${bundle.body[i].beat} at position ${i + 1}).`,
          path: ['body', i, 'beat'],
        });
      }
    }
    // hashtag format: no #, no spaces, lowercased
    for (let i = 0; i < bundle.hashtags.length; i++) {
      const tag = bundle.hashtags[i];
      if (tag.startsWith('#')) {
        ctx.addIssue({
          code: 'custom',
          message: `Hashtag '${tag}' should be stored without the # prefix. The # is added at extraction time.`,
          path: ['hashtags', i],
        });
      }
      if (/\s/.test(tag)) {
        ctx.addIssue({
          code: 'custom',
          message: `Hashtag '${tag}' contains a space.`,
          path: ['hashtags', i],
        });
      }
    }
  });

export type UGCBundle = z.infer<typeof UGCBundleSchema>;

// ============================================================
// Helper getters (mirror the Python @property methods)
// ============================================================

/** Sum of hook + body beats + CTA durations. */
export function totalDurationSeconds(bundle: UGCBundle): number {
  return (
    bundle.hook.duration_seconds +
    bundle.body.reduce((sum, b) => sum + b.duration_seconds, 0) +
    bundle.cta.duration_seconds
  );
}

/** Flat script text for HeyGen / TTS engines. Order: hook → body (beat order) → CTA. */
export function scriptText(bundle: UGCBundle): string {
  const parts: string[] = [bundle.hook.text];
  for (const beat of bundle.body) parts.push(beat.text);
  parts.push(bundle.cta.text);
  return parts.join(' ');
}

File: lib/voice-engine/ugc_schema.py
"""
ugc_schema.py
=============

JSON schema for the UGC (scripted video) output bundle.

The model returns a single structured object containing everything downstream
consumers need:
  - hook                 (0-3s, the attention grab)
  - body                 (3-45s, broken into 1-5 beats)
  - cta                  (3-5s, conversational ask)
  - overlays             (3-8 on-screen text snippets with timing)
  - caption              (post description for the platform)
  - hashtags             (3-5 mix of broad and niche, no # prefix in storage)
  - metadata             (language, platform, swipe-test self-report)

This schema is validated server-side after generation. Validation failures
trigger regeneration with the failure reasons sent back to the model.

Replaces the previous flat {opening, body, closing} JSON. Existing consumers
that called the old shape need to be migrated; ugc_extractor.py handles the
HeyGen flat-string extraction.

Version: 1.0
"""

from __future__ import annotations

import re
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, field_validator


# ============================================================================
# Word-counting helper
# ============================================================================
#
# Hotfix — both word-count validators used `len(v.split())`, which counts
# EVERY whitespace-separated token. That includes pure-emoji tokens like
# ❌ or ✅. So an overlay like "BUFFER ❌ NOTION ❌ FIGMA ❌" (three brand
# names plus three pictogram markers — exactly the "BRAND ❌" pattern
# ugc_validator.py explicitly recommends) was rejected as "6 words; max
# is 5" and the entire video generation failed.
#
# `_count_lexical_words` only counts tokens that contain at least one
# letter or digit (Unicode-aware). Pictogram-only tokens (emoji, dashes,
# bullets, arrows) don't increment the count.
#
#   "BUFFER ❌ NOTION ❌ FIGMA ❌"  → 3 (was 6)
#   "I dropped Buffer last month"  → 5 (unchanged)
#   "$1M in 6 months"              → 4 (unchanged — digits count)
#   "❌"                            → 0 (was 1; pure pictograms aren't
#                                       a word)
#
# Mirrored in lib/voice-engine/ugc-schema.ts (countLexicalWords) so the
# TS and Python paths agree.
_LEXICAL_CHAR_RE = re.compile(r"[^\W_]", re.UNICODE)


def _count_lexical_words(text: str) -> int:
    return sum(1 for token in text.split() if _LEXICAL_CHAR_RE.search(token))


# ============================================================================
# Enums
# ============================================================================

class DeliveryStyle(str, Enum):
    """How the line should be performed on camera or by TTS.

    Used by HeyGen / TTS engines that support delivery hints, and by
    post-production for editing decisions (cut on emphasis, slow on tension).
    """
    PUNCHY = "punchy"               # Fast, high-energy, attention-grabbing
    EXPLANATORY = "explanatory"     # Calm, informative
    TENSION = "tension"             # Building toward a reveal, slightly slower
    REVEAL = "reveal"               # The payoff moment, emphatic but controlled
    WARM = "warm"                   # Friendly, conversational close
    CONFESSIONAL = "confessional"   # First-person admission, slightly vulnerable
    EMPHATIC = "emphatic"           # Stressed, slow. Use sparingly.


# ============================================================================
# Sections
# ============================================================================

class HookSection(BaseModel):
    """The first 0-3 seconds. Single most important piece of the bundle.

    Must pass the 0.5-second swipe test: would a stranger keep watching past
    the first half-second?
    """
    model_config = ConfigDict(extra="forbid")

    text: str = Field(..., min_length=10, max_length=180)
    duration_seconds: float = Field(..., ge=1.0, le=4.0)
    delivery: DeliveryStyle

    @field_validator("text")
    @classmethod
    def hook_word_count(cls, v: str) -> str:
        # Lexical word count (alphanumeric tokens only). See module docstring.
        word_count = _count_lexical_words(v)
        if word_count > 9:
            raise ValueError(
                f"Hook has {word_count} words. Maximum is 9 spoken words. "
                f"Current hook: '{v}'. Trim it aggressively."
            )
        return v


class BodyBeat(BaseModel):
    """One beat of the body section. Beats run sequentially (1, 2, 3...).

    Each beat delivers exactly one idea and ends on tension or specificity
    that pulls the viewer to the next beat.
    """
    model_config = ConfigDict(extra="forbid")

    beat: int = Field(..., ge=1, le=5)
    text: str = Field(..., min_length=20, max_length=400)
    duration_seconds: float = Field(..., ge=2.0, le=15.0)
    delivery: DeliveryStyle


class CTASection(BaseModel):
    """The closing 3-5 seconds. Conversational ask, never a sales pitch."""
    model_config = ConfigDict(extra="forbid")

    text: str = Field(..., min_length=10, max_length=200)
    duration_seconds: float = Field(..., ge=2.0, le=6.0)
    delivery: DeliveryStyle


class Overlay(BaseModel):
    """An on-screen text overlay. Reinforces the spoken word, never repeats it.

    trigger_at_seconds is measured from the start of the video (not the start
    of the section it lands in). Post-production uses this to place text on
    the timeline.
    """
    model_config = ConfigDict(extra="forbid")

    text: str = Field(..., min_length=1, max_length=40)
    trigger_at_seconds: float = Field(..., ge=0.0)
    duration_seconds: float = Field(..., ge=0.5, le=5.0)

    @field_validator("text")
    @classmethod
    def overlay_word_count(cls, v: str) -> str:
        # Lexical word count — emoji like ❌ ✅ → don't inflate the count.
        # ugc_validator.py explicitly recommends "BRAND ❌" shapes;
        # counting the pictogram made that pattern un-shippable.
        word_count = _count_lexical_words(v)
        if word_count > 5:
            raise ValueError(
                f"Overlay '{v}' has {word_count} words; max is 5. "
                f"Overlays longer than 5 words are an anti-pattern."
            )
        return v


class UGCMetadata(BaseModel):
    """Bundle-level metadata. The model is responsible for setting passes_swipe_test
    honestly after running its own check.
    """
    model_config = ConfigDict(extra="forbid")

    language: str = Field(default="en", min_length=2, max_length=8)
    platform: str  # tiktok, instagram, threads, linkedin, facebook
    passes_swipe_test: bool = Field(
        default=True,
        description="Self-reported by the model after running the 0.5-second swipe test.",
    )


# ============================================================================
# UGCBundle (top-level)
# ============================================================================

class UGCBundle(BaseModel):
    """The complete UGC output. Single source of truth.

    Downstream consumers:
      - HeyGen / TTS:        ugc_extractor.extract_script_for_heygen(bundle)
      - Video editor:        ugc_extractor.extract_overlay_track(bundle)
      - Social scheduler:    ugc_extractor.extract_caption_for_post(bundle)
      - Storyboard view:     ugc_extractor.extract_beat_breakdown(bundle)
    """
    model_config = ConfigDict(extra="forbid")

    hook: HookSection
    body: list[BodyBeat] = Field(..., min_length=1, max_length=5)
    cta: CTASection
    overlays: list[Overlay] = Field(default_factory=list, min_length=3, max_length=8)
    caption: str = Field(..., min_length=20, max_length=500)
    hashtags: list[str] = Field(default_factory=list, min_length=3, max_length=5)
    metadata: UGCMetadata

    @property
    def total_duration_seconds(self) -> float:
        """Sum of hook + all body beats + cta. Used for validation."""
        return (
            self.hook.duration_seconds
            + sum(b.duration_seconds for b in self.body)
            + self.cta.duration_seconds
        )

    @property
    def script_text(self) -> str:
        """Flat concatenated script for HeyGen / TTS engines.

        Order: hook + body beats (in beat order) + cta. Spaces between sections.
        """
        parts = [self.hook.text]
        parts.extend(b.text for b in self.body)
        parts.append(self.cta.text)
        return " ".join(parts)

    @field_validator("body")
    @classmethod
    def body_beats_sequential(cls, v: list[BodyBeat]) -> list[BodyBeat]:
        """Beats must be numbered 1, 2, 3... in order. Catches model errors
        where it skips a beat number or ships them out of order.
        """
        for i, beat in enumerate(v, 1):
            if beat.beat != i:
                raise ValueError(
                    f"Body beats must be sequential starting from 1 "
                    f"(got beat={beat.beat} at position {i})."
                )
        return v

    @field_validator("hashtags")
    @classmethod
    def hashtags_format(cls, v: list[str]) -> list[str]:
        """Hashtags stored without the # prefix; spaces not allowed."""
        cleaned: list[str] = []
        for tag in v:
            if tag.startswith("#"):
                raise ValueError(
                    f"Hashtag '{tag}' should be stored without the # prefix. "
                    f"The # is added at extraction time."
                )
            if " " in tag:
                raise ValueError(f"Hashtag '{tag}' contains a space.")
            cleaned.append(tag.lower().strip())
        return cleaned
